### PR TITLE
Add calculate element address with element stride

### PR DESCRIPTION
**Pull request by a7ehuo**
Update `TransformUtil` to calculate element address with element stride and add `flattenedArrayElementSize` to `ClassEnv`.

### CODE REVIEW

**Comment by a7ehuo** [comment]
@hzongaro @0xdaryl, may I ask you to review? Thanks!
@mpirvu, may I ask you to review the change related to JITServer? Thanks!

**Review by mpirvu — COMMENTED** [review]
LGTM from JITServer point of view. I don't see `flattenedArrayElementSize()` is used though.
As an optimization, in the future we could keep the desired size in the cache of J9classes at the server.

**Comment by a7ehuo** [comment]
> I don't see flattenedArrayElementSize() is used though.

`flattenedArrayElementSize` will be used in a future change that will transform flattened array element load/store helper calls with sym refs in VP.

File: runtime/compiler/optimizer/J9TransformUtil.hpp [hunk 1]
```diff
@@ -201,8 +201,23 @@ class OMR_EXTENSIBLE TransformUtil : public OMR::TransformUtilConnector
...
-   static TR::Node *calculateElementAddress(TR::Compilation *, TR::Node *array, TR::Node * index, TR::DataType type);
+   /**
+    * \brief
+    *    Calculate offset for an array element given its index based on `type` or `elementStride`
+    *
+    * \param index
+    *    The index into the array
+    *
+    * \param type
+    *    The data type of the array element. If `elementStride` is none zero, `type` is ignored
```
Comment (hzongaro) [inline runtime/compiler/optimizer/J9TransformUtil.hpp:212]: Typo - "none zero" should be "nonzero"

File: runtime/compiler/optimizer/J9TransformUtil.hpp [hunk 2]
```diff
@@ -201,8 +201,23 @@ class OMR_EXTENSIBLE TransformUtil : public OMR::TransformUtilConnector
...
+    *
+    * \param elementStride
+    *    The size of the array element. If it is zero, the width is calculated based on TR::DataType `type`
+    *
+    * \return The Int64 value representing the offset into an array of the specified type given the index
+    */
+   static TR::Node *calculateOffsetFromIndexInContiguousArray(TR::Compilation *, TR::Node * index, TR::DataType type, int32_t elementStride = 0);
```
Comment (hzongaro) [inline runtime/compiler/optimizer/J9TransformUtil.hpp:219]: Zero might be a valid element stride for an array of a value type that has no fields.  I'm just verifying that.  If I'm correct, you'll need to use a different way of signalling that the `type` parameter should be used.

One possibility might be to split this into two methods - one that accepts only the `type` parameter, and one that accepts only the `elementStride` parameter, and then the version that uses the `type` parameter could call the version that uses `elementStride`.

Another possibility might be to use a negative stride.

File: runtime/compiler/optimizer/J9TransformUtil.cpp [hunk 3]
```diff
@@ -1670,13 +1670,22 @@ TR::Node * J9::TransformUtil::calculateOffsetFromIndexInContiguousArray(TR::Comp
...
       offset = TR::Node::create(shlOp, 2, offset, shiftNode);
       }
+
+   if (elementStride != 0)
+      {
+      TR::Node *elementStrideNode = TR::Node::create(constOp, 0);
+      elementStrideNode->setConstValue(elementStride);
+      offset = TR::Node::create(mulOp, 2, offset, elementStrideNode);
+      }
+
```
Comment (hzongaro) [inline runtime/compiler/optimizer/J9TransformUtil.cpp:1688]: You might want to consider checking whether the stride is a power of two and using a shift if it is.

**Review by hzongaro — COMMENTED** [review]
The general approach looks correct - we just need to check on the possibility that strides of length zero might occur

File: runtime/compiler/optimizer/J9TransformUtil.hpp (hunk 2 above)
Comment (a7ehuo) [reply runtime/compiler/optimizer/J9TransformUtil.hpp:219]: Thanks for the suggestion! I'll split them into two APIs:

One takes `TR::DataType type`
```
   static TR::Node *calculateOffsetFromIndexInContiguousArray(TR::Compilation *, TR::Node * index, TR::DataType type);
   static TR::Node *calculateElementAddress(TR::Compilation *, TR::Node *array, TR::Node * index, TR::DataType type);
```
Another takes `int32_t elementStride`
```
   static TR::Node *calculateOffsetFromIndexInContiguousArrayWithElementStride(TR::Compilation *, TR::Node * index, int32_t elementStride);
   static TR::Node *calculateElementAddressWithElementStride(TR::Compilation *, TR::Node *array, TR::Node * index, int32_t elementStride);
```

File: runtime/compiler/optimizer/J9TransformUtil.hpp (hunk 1 above)
Comment (a7ehuo) [reply runtime/compiler/optimizer/J9TransformUtil.hpp]: The comments are updated in cf9ae54

File: runtime/compiler/optimizer/J9TransformUtil.hpp (hunk 2 above)
Comment (a7ehuo) [reply runtime/compiler/optimizer/J9TransformUtil.hpp]: Addressed in cf9ae54

File: runtime/compiler/optimizer/J9TransformUtil.cpp (hunk 3 above)
Comment (a7ehuo) [reply runtime/compiler/optimizer/J9TransformUtil.cpp]: Addressed in cf9ae54

File: runtime/compiler/optimizer/J9TransformUtil.hpp (hunk 2 above)
Comment (a7ehuo) [reply runtime/compiler/optimizer/J9TransformUtil.hpp]: I noticed an issue with the refactored code in 7bfe3e8. `_widthToShift[]` has size 9 but `elementStride` could be any value

```
   if (elementStride && (elementStride % 2 == 0))
       shift = TR::TransformUtil::convertWidthToShift(elementStride);
``

Comment (a7ehuo) [reply runtime/compiler/optimizer/J9TransformUtil.hpp]: I have a fix now. Change the status to `WIP` while I'm running more tests.

**Review by hzongaro — APPROVED** [review]
I think the changes look good.  Thanks!

**Comment by a7ehuo** [comment]
@0xdaryl All comments are addressed. Ready for another review. Thanks!

**Comment by 0xdaryl** [comment]
Jenkins build all

**Comment by 0xdaryl** [comment]
Jenkins build all

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk17

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**